### PR TITLE
Ecarton/cumulus 3751 granule collection update

### PR DESCRIPTION
**Pull request by etcart**
**Summary:** allow configuration of moveGranules task with new collection in order to move a granule to the given collection, along with its files

Addresses [CUMULUS-3751: Develop amazing new feature](https://bugs.earthdata.nasa.gov/browse/CUMULUS-3751)

## Changes

* moveGranules source bucket is no longer configured by config.bucket param, but by granule files configured bucket
* moveGranules target information is still configured by collection, but with an acceptance of "different" collection indicating a need to update granule in datastore and workflow, along with movement of files to target collection

## PR Checklist

- [ ] Update CHANGELOG
- [ ] Unit tests
- [ ] Ad-hoc testing - Deploy changes and test manually
- [ ] Integration tests

### CODE REVIEW

File: tasks/move-granules/index.js [hunk 1]
```diff
@@ -49,28 +53,83 @@ function buildGranuleDuplicatesObject(movedGranulesByGranuleId) {
...
+ * @param {ApiGranuleRecord} granule
+ * @param {CollectionRecord} collection
+ * @returns {}
+ */
+async function updateGranuleCollection(granule, collection) {
+  const collectionId = constructCollectionId(collection.name, collection.version);
+  const updatedGranule = { ...granule };
+  if (granule.collectionId !== collectionId) {
+    updatedGranule.collectionId = collectionId;
+    // TODO api call to update granule in pg here
```
Comment (npauzenga) [inline tasks/move-granules/index.js:67]: Zooming out a bit here. Is an API call what we want here? This task could be run on x number of granules. The API call to update a granule is likely going to be quick but this task could result in e.g. 1k invocations to update all of the granules*.

Why use an API write as opposed to a workflow message write? In other words, is there anything we can do to update the granules in the message payload with the new collection info and then let the standard workflow writes handle updating the datastores? 

That may not be feasible for some reason. It would also still require updating the granule upsert logic. What it would get us (theoretically) is a built-in mechanism for updating several granules at once.

*But of course we can instruct users to use this in a way that makes sense. e.g. small batches of granules.

File: tasks/move-granules/index.js [hunk 2]
```diff
@@ -144,13 +187,12 @@ async function updateGranuleMetadata(granulesObject, collection, cmrFiles, bucke
...
 async function moveFileRequest(
   file,
-  sourceBucket,
   duplicateHandling,
   markDuplicates = true,
   s3MultipartChunksizeMb
 ) {
   const source = {
-    Bucket: sourceBucket,
+    Bucket: file.sourceBucket,
```
Comment (npauzenga) [inline tasks/move-granules/index.js:195]: Is there any use case where this bites us? I suppose it would need to be a situation where the file has a bucket that isn't what the task wants. So if the file has `sourceBucket1` but the task is passed `sourceBucket2` as almost an override...

I don't know if that's a real thing. Just that move-granules is used quite a lot in ingest workflows so wanna be sure we're covering all our bases.

Comment (etcart) [reply tasks/move-granules/index.js:195]: yea, I spent a lot of time thinking about this. it's possible that we want to allow config.bucket as an override, but that has the problem that in *most* cases a granule that is already where it belongs will have files in *multiple* buckets. the case where there is "one master bucket" that all files are stored in is a special case for moving from staging to final location.
but in that case it looks like the granule should have its files correctly written with the staging bucket configured as the file's bucket attribute. maybe I need to follow up to ensure that that's true but that's why it's listed as a breaking change in the changeling

File: example/spec/parallel/moveGranules/MoveGranulesCollectionSpec.js [hunk 3]
```diff
@@ -0,0 +1,174 @@
+'use strict';
+
+// TODO: rename this file to be clear it is also testing moving zero byte files
+
+/**
+ * LocalStack has a bug where it ignores the `Tagging` parameter to the
```
Comment (Jkovarik) [inline example/spec/parallel/moveGranules/MoveGranulesCollectionSpec.js:6]: Is this copypasta or sometihng new?

File: tasks/move-granules/index.js [hunk 4]
```diff
@@ -25,13 +25,16 @@ const {
...
-
 const { urlPathTemplate } = require('@cumulus/ingest/url-path-template');
 const { isFileExtensionMatched } = require('@cumulus/message/utils');
 const log = require('@cumulus/common/log');
+const { constructCollectionId } = require('@cumulus/message/Collections');
 
 const MB = 1024 * 1024;
-
+/**
+* @typedef { import('@cumulus/types').ApiGranuleRecord } ApiGranuleRecord
```
Comment (Jkovarik) [inline tasks/move-granules/index.js:35]: I don't think this type is right - an ApiGranuleRecord, as it's not for sure going to have some of the required keys.

File: tasks/move-granules/index.js (hunk 1 above)
Comment (Jkovarik) [reply tasks/move-granules/index.js:67]: Definitely a good follow-on, @Nnaga1 / you should nail down the reporting approach, possibly have an arch session with the team

File: example/spec/parallel/moveGranules/MoveGranulesCollectionSpec.js (hunk 3 above)
Comment (etcart) [reply example/spec/parallel/moveGranules/MoveGranulesCollectionSpec.js:6]: this started life at a copy of MoveGranulesTagsSpec.js
I could probably afford to follow up on that TODO in the tags side and rearrange it in collection

**Comment by npauzenga** [comment]
@etcart I'm removing the Needs Review label because this is still in draft and I think there was some re-work happening? If that's incorrect, feel free to re-add the label, update the branch, etc.
